ipl/main.c

```diff
@@ -1929,6 +1940,12 @@ int fix_attributes(char *path, u32 *total)
 			if (res != FR_OK || fno.fname[0] == 0)
 				break;
 
+			if (is_root && !strcmp(fno.fname, "Nintendo"))
```

Comment:
Not above `i = strlen(path);`
This breaks the whole path (invalidates it after Nintendo folder if i==0).

It should be like this:
```
// Set new directory.
i = strlen(path);
if (is_root && !strcmp(fno.fname, "Nintendo"))
{
	path[i] = 0;
	continue;
}
memcpy(&path[i], "/", 1);
```

Reply:
That's a good point. I think we can move that `i = strlen(path);` line even outside the for loop. We don't need to recalculate the parent directory with every file.

Another question:
```
for (k = 0; k < 256; k++)
{
	if (fno.fname[k] == 0)
		break;
}
```

How is this different than strlen? is there some hidden time-saving thing that I'm missing?

Reply:
I think `path[dirLength + fileLength + 2] = 0;` is not needed either. The null character will be captured from the previous line. I'll test all these theories to be sure they are valid.

Reply:
You can't move `i = strlen(path);` outside of the loop, it will break the enter directory code.

The loop is indeed strlen. It can be reduced to `k = strlen(fno.fname)`

The `path[dirLength + fileLength + 2] = 0;` is not needed, yes.
fname is null terminated and taken from the +1 in `memcpy(&path[i+1], fno.fname, k + 1);`

So, more optimizations to do.

Reply:
Btw, k is not needed. It can be removed and the code from
```
for (k = 0; k < 256; k++)
{
	if (fno.fname[k] == 0)
		break;
}
memcpy(&path[i+1], fno.fname, k + 1);
path[i + k + 2] = 0;
```
Can become:
```
memcpy(&path[i+1], fno.fname, strlen(fno.fname) + 1);
```